File: DSA_python/DSA11008 KIỂM TRA NODE LÁ.py
from collections import deque

class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right


def add_to_tree(edges):
    if not edges:
        return

    nodes = dict()

    for parent, child, pos in edges:
        if parent not in nodes:
            nodes[parent] = TreeNode(parent)
        if child not in nodes:
            nodes[child] = TreeNode(child)
        if pos == 'L':
            nodes[parent].left = nodes[child]
        else:
            nodes[parent].right = nodes[child]
        
    children = [child for _,child,_ in edges]
    parents = [parent for parent,_,_ in edges]
    root = parents[0]

    for parent in parents:
        if parent not in children:
            root = parent
    
    return nodes[root]

def traversal(node, depth, mp):
    if not node:
        mp.add(depth)
        return

    traversal(node.left, depth+1, mp)
    traversal(node.right, depth+1, mp)

def submit():
    n = int(input())
    A = list(map(str, input().split()))
    edges = list()

    for i in range(2, len(A), 3):
        edges.append((A[i-2], A[i-1], A[i]))

    root = add_to_tree(edges)
    mp = set()
    traversal(root, 0, mp)

    print(1 if len(mp) == 1 else 0)







test = int(input())
for i in range(test):
    submit()

File: DSA/DSA11008 KIỂM TRA NODE LÁ.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};



bool add(TreeNode *node, int parent, int child, char pos)
{
    if (node == nullptr)
    {
        return false;
    }
    
    if (node->val != parent)
    {
        if (!add(node->left, parent, child, pos))
        {
            add(node->right, parent, child, pos);
        }
    }
    else
    {
        if (pos == 'L')
        {
            node->left = new TreeNode(child);
        }
        else
        {
            node->right = new TreeNode(child);
        }
        return true;
    }
    return false;
}

void lvl_traversal(TreeNode *node, int depth, unordered_set<int> &ans)
{
    if (node == nullptr)
    {
        ans.insert(depth);
        return;
    }
    
    lvl_traversal(node->left, depth+1, ans);
    lvl_traversal(node->right, depth+1, ans);
}

void submit()
{
    TreeNode* root = nullptr;

    int n;
    cin >> n;

    int parent;
    int child;
    char pos;
    cin >> parent >> child >> pos;

    root = new TreeNode(parent);
    
    add(root, parent, child, pos);

    
    for (int i = 1; i < n; i++)
    {
        cin >> parent >> child >> pos;
        
        add(root, parent, child, pos);
    }
    
    unordered_set<int> ans;
    lvl_traversal(root, 0, ans);

    ans.size() == 1 ? cout << 1 : cout << 0;
    


    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}